# Load and Merge 

In this notebook we will be loading the Freddic Mac data Single Family Loan dataset origination and performance, assigning columns names to origination and we will be merging the performance files too.

# Output
- Merged dataset saved to data/processed/master.pkl
- One loan per row with DEFAULT_FLAG column added

In [1]:
import pandas as pd
import numpy as np 
from src.features import ORIG_COLS,KEEP_COLS,MISSING_CODES
from pathlib import Path

In [2]:
data_dir = Path("../data")

test_df = pd.read_csv(
    data_dir / "sample_2000" / "sample_orig_2000.txt",
    sep="|",
    header=None,
    low_memory=False
)

In [3]:
#Looking at the first row as columns are unnamed
print("Shape:", test_df.shape)
print("\nFirst row as dict:")
for idx, val in test_df.iloc[0].items():
    print(f"  col {idx}: {val}")

Shape: (50000, 32)

First row as dict:
  col 0: 790
  col 1: 200211
  col 2: N
  col 3: 203003
  col 4: nan
  col 5: 30
  col 6: 1
  col 7: P
  col 8: 92
  col 9: 29
  col 10: 101000
  col 11: 92
  col 12: 6.125
  col 13: R
  col 14: N
  col 15: FRM
  col 16: WI
  col 17: SF
  col 18: 53100
  col 19: F00Q10000035
  col 20: P
  col 21: 329
  col 22: 2
  col 23: Other sellers
  col 24: Other servicers
  col 25: nan
  col 26: nan
  col 27: 9
  col 28: nan
  col 29: 7
  col 30: N
  col 31: 9


In [4]:
#imported Loader Function
from src.loader import orig_loader


In [5]:
df = orig_loader()

In [6]:
df.shape

(550000, 33)

In [7]:
df.columns

Index(['CREDIT_SCORE', 'FIRST_PYMT_DATE', 'FIRST_TIME_BUYER', 'MATURITY_DATE',
       'MSA', 'MI_PCT', 'NUM_UNITS', 'OCC_STATUS', 'OCLTV', 'DTI', 'UPB',
       'LTV', 'INTEREST_RATE', 'CHANNEL', 'PPM_FLAG', 'PRODUCT_TYPE',
       'PROPERTY_STATE', 'PROPERTY_TYPE', 'ZIP_3', 'LOAN_SEQ_NUM',
       'LOAN_PURPOSE', 'LOAN_TERM', 'NUM_BORROWERS', 'SELLER_NAME',
       'SERVICER_NAME', 'SUPER_CONF_FLAG', 'PRE_HARP_LOAN_SEQ',
       'SPECIAL_ELIGIBILITY', 'RELIEF_REFINANCE_FLAG', 'PROPERTY_VAL_METHOD',
       'INTEREST_ONLY_FLAG', 'MI_CANCELLATION_FLAG', 'ORIG_YEAR'],
      dtype='str')

In [8]:
df.iloc[0]

CREDIT_SCORE                         790
FIRST_PYMT_DATE                   200211
FIRST_TIME_BUYER                       N
MATURITY_DATE                     203003
MSA                                  NaN
MI_PCT                                30
NUM_UNITS                              1
OCC_STATUS                             P
OCLTV                                 92
DTI                                   29
UPB                               101000
LTV                                   92
INTEREST_RATE                      6.125
CHANNEL                                R
PPM_FLAG                               N
PRODUCT_TYPE                         FRM
PROPERTY_STATE                        WI
PROPERTY_TYPE                         SF
ZIP_3                              53100
LOAN_SEQ_NUM                F00Q10000035
LOAN_PURPOSE                           P
LOAN_TERM                            329
NUM_BORROWERS                          2
SELLER_NAME                Other sellers
SERVICER_NAME   

In [9]:
df.isnull().sum()

CREDIT_SCORE                  0
FIRST_PYMT_DATE               0
FIRST_TIME_BUYER              0
MATURITY_DATE                 0
MSA                      173697
MI_PCT                        0
NUM_UNITS                     0
OCC_STATUS                    0
OCLTV                         0
DTI                           0
UPB                           0
LTV                           0
INTEREST_RATE                 0
CHANNEL                       0
PPM_FLAG                      0
PRODUCT_TYPE                  0
PROPERTY_STATE                0
PROPERTY_TYPE                 0
ZIP_3                         0
LOAN_SEQ_NUM                  0
LOAN_PURPOSE                  0
LOAN_TERM                     0
NUM_BORROWERS                 0
SELLER_NAME                   0
SERVICER_NAME                 0
SUPER_CONF_FLAG          547869
PRE_HARP_LOAN_SEQ        529995
SPECIAL_ELIGIBILITY           0
RELIEF_REFINANCE_FLAG    529995
PROPERTY_VAL_METHOD           0
INTEREST_ONLY_FLAG            0
MI_CANCE

# Note 
The Freddic Mac data also contains null valus as form of integer/codes i.e. Credit Score :- 9999 refers to Not available we will take care of it later though 

In [10]:
df.nunique()

CREDIT_SCORE                397
FIRST_PYMT_DATE             160
FIRST_TIME_BUYER              3
MATURITY_DATE               448
MSA                         455
MI_PCT                       40
NUM_UNITS                     5
OCC_STATUS                    3
OCLTV                       175
DTI                          66
UPB                         720
LTV                         121
INTEREST_RATE              1024
CHANNEL                       5
PPM_FLAG                      2
PRODUCT_TYPE                  1
PROPERTY_STATE               54
PROPERTY_TYPE                 6
ZIP_3                       892
LOAN_SEQ_NUM             550000
LOAN_PURPOSE                  3
LOAN_TERM                   241
NUM_BORROWERS                 3
SELLER_NAME                  42
SERVICER_NAME                42
SUPER_CONF_FLAG               1
PRE_HARP_LOAN_SEQ         20005
SPECIAL_ELIGIBILITY           1
RELIEF_REFINANCE_FLAG         1
PROPERTY_VAL_METHOD           1
INTEREST_ONLY_FLAG            1
MI_CANCE

In [11]:
print(df.dtypes)

CREDIT_SCORE               int64
FIRST_PYMT_DATE            int64
FIRST_TIME_BUYER             str
MATURITY_DATE              int64
MSA                      float64
MI_PCT                     int64
NUM_UNITS                  int64
OCC_STATUS                   str
OCLTV                      int64
DTI                        int64
UPB                        int64
LTV                        int64
INTEREST_RATE            float64
CHANNEL                      str
PPM_FLAG                     str
PRODUCT_TYPE                 str
PROPERTY_STATE               str
PROPERTY_TYPE                str
ZIP_3                      int64
LOAN_SEQ_NUM                 str
LOAN_PURPOSE                 str
LOAN_TERM                  int64
NUM_BORROWERS              int64
SELLER_NAME                  str
SERVICER_NAME                str
SUPER_CONF_FLAG           object
PRE_HARP_LOAN_SEQ         object
SPECIAL_ELIGIBILITY        int64
RELIEF_REFINANCE_FLAG     object
PROPERTY_VAL_METHOD        int64
INTEREST_O

# Note 

Some Column like FIRST_PYMT_DATE contains date as int64 so we may need to transform them 

# Deciding Columns to Keep Based on Their Description

| Field | Decision | Why |
|-------|----------|-----|
| **CREDIT_SCORE** | Keep | FICO score of the borrower at the time the loan was issued (300–850). This is the single most important indicator of borrower quality. |
| **FIRST_PYMT_DATE** | Keep | Date of the first mortgage payment in `YYYYMM` format. Required for quarterly analysis. |
| **FIRST_TIME_BUYER** | Keep | First-time buyers during the housing bubble were often less financially stable. |
| **MATURITY_DATE** | Drop | Redundant with **LOAN_TERM**. |
| **MSA** | Drop | Metropolitan Statistical Area code. Contains 455 unique values and 173,697 null values (~31% of the data). Geographic information is useful, but this is largely redundant because **PROPERTY_STATE** is available. |
| **MI_PCT** | Keep | Mortgage insurance percentage. This is the lender's own risk signal—if mortgage insurance was required, the lender already considered the loan riskier. |
| **NUM_UNITS** | Keep | Number of units in the property. Multi-unit properties are often investor purchases, and investors tend to walk away from loans more frequently than owner-occupants. |
| **OCC_STATUS** | Keep *(May Drop Later)* | Occupancy status (Primary Residence, Investor Property, Second Home). Since investor properties are generally riskier, this may be useful. It may later prove redundant with **NUM_UNITS**. |
| **OCLTV** | Keep | Original Combined Loan-to-Value ratio. Includes both the primary mortgage and any secondary loans. It contains more information than **LTV**. For example, a borrower could have **LTV = 80** but **OCLTV = 100** due to a second mortgage. Higher OCLTV indicates higher risk. |
| **DTI** | Keep | Debt-to-Income ratio. Total monthly debt payments divided by total monthly income at origination. Indicates how financially stretched the borrower was. |
| **UPB** | Keep | Unpaid Principal Balance. Larger loans generally carry greater absolute financial risk. |
| **LTV** | Keep | Loan-to-Value ratio. Although it contains less information than **OCLTV**, it is still a valuable measure of mortgage risk. |
| **INTEREST_RATE** | Keep | Interest rate at loan origination. Higher rates generally indicate that the lender priced in greater borrower risk. |
| **CHANNEL** | Keep | Indicates how the loan was originated. Broker-originated loans tended to be riskier because brokers often had little incentive to verify borrower income or assets thoroughly. |
| **PPM_FLAG** | Drop | Indicates whether the borrower is penalized for paying off the loan early. This is a very weak predictor since borrowers who default generally do not repay the loan anyway. |
| **PRODUCT_TYPE** | Drop | Indicates Fixed-Rate or Adjustable-Rate mortgages. The dataset contains only one unique value, resulting in zero variance and no predictive information. |
| **PROPERTY_STATE** | Keep | Two-letter state abbreviation representing the property's location. Geographic location is an important factor in mortgage performance. |
| **PROPERTY_TYPE** | Keep | Different property types exhibit different default rates. |
| **ZIP_3** | Drop | First three digits of the ZIP code. Contains 892 unique values, resulting in high cardinality. It is also largely redundant with **PROPERTY_STATE**. |
| **LOAN_SEQ_NUM** | Keep *(For Joining Only)* | Unique identifier for each loan. Used only for joining datasets. |
| **LOAN_PURPOSE** | Keep | Purpose of the loan (e.g., Purchase, Cash-out Refinance, etc.). |
| **LOAN_TERM** | Keep | Number of scheduled monthly payments. May contain non-standard values and could require transformation. |
| **NUM_BORROWERS** | Keep | Having a co-borrower provides an additional income source, potentially reducing default risk. |
| **SELLER_NAME** | Drop | Name of the entity that sold the mortgage to Freddie Mac. Not required for our analysis. |
| **SERVICER_NAME** | Drop | Name of the loan servicer as of the last reported period. Not required for our analysis. |
| **SUPER_CONF_FLAG** | Drop | Marks loans above conforming loan limits. Applicable only from October 2008 onward and contains nearly 99% null values. |
| **PRE_HARP_LOAN_SEQ** | Drop | Links loans refinanced under HARP. Since HARP was introduced in 2009, this field contains nearly 99% null values. |
| **SPECIAL_ELIGIBILITY** | Drop | Identifies loans under Freddie Mac special programs (e.g., Home Possible). Available only after March 2015. Contains zero variance, with all values marked as "Not Available." |
| **RELIEF_REFINANCE_FLAG** | Drop | Indicates HARP relief refinance loans. Since this is a post-2008 program, approximately 96% of the values are null. |
| **PROPERTY_VAL_METHOD** | Drop | Method used to value the property. Available only after 2017 and contains only one unique value (zero variance). |
| **INTEREST_ONLY_FLAG** | Drop | Indicates whether the borrower initially paid only interest without reducing principal. Contains only one unique value across the dataset. |
| **MI_CANCELLATION_FLAG** | Drop | Indicates whether mortgage insurance was cancelled after Freddie Mac purchased the loan. This is a post-origination event and has zero variance. |
| **ORIG_YEAR** | Keep | Year extracted by our `orig_loader()` function. Useful for time-based analysis. |

In [12]:
df = df[KEEP_COLS]

In [13]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 550000 entries, 0 to 549999
Data columns (total 19 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   CREDIT_SCORE      550000 non-null  int64  
 1   FIRST_PYMT_DATE   550000 non-null  int64  
 2   FIRST_TIME_BUYER  550000 non-null  str    
 3   MI_PCT            550000 non-null  int64  
 4   NUM_UNITS         550000 non-null  int64  
 5   OCC_STATUS        550000 non-null  str    
 6   OCLTV             550000 non-null  int64  
 7   DTI               550000 non-null  int64  
 8   UPB               550000 non-null  int64  
 9   LTV               550000 non-null  int64  
 10  INTEREST_RATE     550000 non-null  float64
 11  CHANNEL           550000 non-null  str    
 12  PROPERTY_STATE    550000 non-null  str    
 13  PROPERTY_TYPE     550000 non-null  str    
 14  LOAN_SEQ_NUM      550000 non-null  str    
 15  LOAN_PURPOSE      550000 non-null  str    
 16  LOAN_TERM         550000 non-nu

In [14]:
# Required NAN Trasfromation in some fields as mentioned above 

for col, code in MISSING_CODES.items():
    df[col] = df[col].replace(code,np.nan)


In [15]:
# check nulls appeared where 9999/999/99 were
print(df[["CREDIT_SCORE", "DTI", "OCLTV", 
            "MI_PCT", "NUM_UNITS", "NUM_BORROWERS"]].isnull().sum())

CREDIT_SCORE      1594
DTI              29998
OCLTV               33
MI_PCT              12
NUM_UNITS            9
NUM_BORROWERS      166
dtype: int64


# Note
Hidden missing codes (9999, 999, 99) replaced with NaN.
Actual handling dropping or imputing will be done 
in notebook 02 during feature engineering.

# Performance File — Columns We Are Loading

| Field | Decision | Why |
|-------|----------|-----|
| **LOAN_SEQ_NUM** | Keep | This is the unique loan identifier that links each performance record back to the origination data. Without this we cannot join the two files. |
| **DELINQ_STATUS** | Keep | This tracks how many months past due the borrower is in each monthly record. A value of 3 or more means the borrower is 90 or more days past due which is the standard definition of serious delinquency and what we are using as our default threshold. |
| **ZERO_BAL_CODE** | Keep | This tells us why the loan balance went to zero. We are only interested in two codes. Code 03 means the loan ended in a short sale or charge off where the lender accepted a loss. Code 09 means the bank foreclosed on the property, took ownership and sold it themselves which is the classic REO outcome. For most loans this field will be null because the loan never terminated during our observation window. |

In [16]:
from src.loader import perf_loader

perf = perf_loader()

print("Shape:", perf.shape)
print("\nSample:")
print(perf.head())
print("\nColumn types:")
print(perf.dtypes)

Shape: (33058391, 4)

Sample:
   LOAN_SEQ_NUM DELINQ_STATUS  ZERO_BAL_CODE  PERF_YEAR
0  F00Q10000035             0            NaN       2000
1  F00Q10000035             0            NaN       2000
2  F00Q10000035             0            NaN       2000
3  F00Q10000035             0            NaN       2000
4  F00Q10000035             0            NaN       2000

Column types:
LOAN_SEQ_NUM         str
DELINQ_STATUS        str
ZERO_BAL_CODE    float64
PERF_YEAR          int64
dtype: object


As we can perf contains monthly data we neeed to handle it first and then merge 

In [17]:
def collapse_perf(perf):
    """
    Collapse performance data from one row per loan per month
    to one row per loan.
    We take the worst value of each column across all months
    so we capture if a loan ever reached serious delinquency
    or ever got foreclosed at any point in its history.
    """
    perf_collapsed = perf.groupby("LOAN_SEQ_NUM").agg(
        MAX_DELINQ  = ("DELINQ_STATUS", "max"), # worst delinquency the loan ever reached
        ZERO_BAL    = ("ZERO_BAL_CODE", "max"), # foreclosure code if it ever happened
        PERF_MONTHS = ("PERF_YEAR", "count")    # total months of performance history
    ).reset_index()

    return perf_collapsed

# Transformer features as folllow
- DELINQ_STATUS->MAX_DELINQ # worst delinquency the loan ever reached
- ZERO_BAL_CODE->ZERO_BAL # foreclosure code if it ever happened
- PERF_YEAR->PERF_MONTHS # total months of performance history


In [18]:
perf = collapse_perf(perf)

In [19]:
df = df.merge(perf, on="LOAN_SEQ_NUM",how='left')

In [20]:
df.head()

,CREDIT_SCORE,FIRST_PYMT_DATE,FIRST_TIME_BUYER,MI_PCT,NUM_UNITS,OCC_STATUS,OCLTV,DTI,UPB,LTV,...,PROPERTY_STATE,PROPERTY_TYPE,LOAN_SEQ_NUM,LOAN_PURPOSE,LOAN_TERM,NUM_BORROWERS,ORIG_YEAR,MAX_DELINQ,ZERO_BAL,PERF_MONTHS
0,790.0,200211,N,30.0,1.0,P,92.0,29.0,101000,92,...,WI,SF,F00Q10000035,P,329,2.0,2000,0,1.0,10
1,771.0,200011,N,0.0,1.0,P,61.0,30.0,50000,61,...,IL,SF,F00Q10000049,C,180,1.0,2000,0,1.0,12
2,762.0,200110,N,0.0,3.0,P,76.0,23.0,90000,76,...,RI,SF,F00Q10000054,N,342,2.0,2000,0,1.0,4
3,737.0,200104,N,25.0,1.0,P,87.0,35.0,170000,87,...,IL,CO,F00Q10000090,P,349,1.0,2000,0,1.0,9
4,594.0,200003,N,0.0,1.0,P,80.0,24.0,49000,80,...,AL,SF,F00Q10000095,N,360,1.0,2000,0,1.0,23


In [21]:
df.shape

(550000, 22)

In [24]:

df.to_pickle("../data/processed/master.pkl")


# Conclusion

Merged orig and perf succesfully and named columns and join them via LOAN_SEQ_NUM and exposrted as pkl file